Demo the concept of the TDD process

ctwdtw opened the pull request:
Here is my understanding of the use case.  I only implement it partially to demo the concept of the TDD process.

#### Routing through Universal Link Use Case

Data:
	- Link Pattern 
	- Link URL
	
Primary course:
	1. System defines routing destination for the above Link Pattern
	2. User initiates a routing through the above Link URL 
	3. System extract data from the Link URL
	4. System display the extracted data on a pre-defined destination

Failed to extract - error course
	1. System displays an error message

ctwdtw committed:
add gitignore

ctwdtw committed:
pod install

ctwdtw committed:
fix code signing build error

ctwdtw committed:
rename `Router` to `AppRouter`

ctwdtw committed:
we want to test the interaction between our `AppRouter` and the `Navigator` of 3rd party, therefore we use the protocol-based mock strategy.

ctwdtw committed:
`AppRouter` asks `Navigator` to register desired link pattern when `AppRouter` is inited

ctwdtw committed:
`AppRouter` request `Navigator` maps extracted value to `CategoryVC` when `Navigator` was trigged to extract Int value `id` from link

ctwdtw committed:
`AppRouter` request `Navigator` fallback mappings to `ErrorVC` when `Navigator` failed to extract `id`

the-yuyut commented on AppRouterTests/AppRouterTests.swift:
```diff
@@ -9,56 +9,106 @@ import XCTest
+        let navigatorSpy = NavigatorSpy()
+        let sut = AppRouter(navigator: navigatorSpy)
+        trackForMemoryLeaks(sut)
+        trackForMemoryLeaks(navigatorSpy)
```
TrackForMemoryLeaks can show Failure message inline with caller, make it more readable.
```suggestion
        trackForMemoryLeaks(sut, file: file, line: line)
        trackForMemoryLeaks(navigatorSpy, file: file, line: line)
```

ctwdtw committed:
Rename helper function from `triggeredToExtract(_:)` to `mappingVC(from:)` for better expressing that
the spy is mimicking an action of mapping extracted values from user triggered `URL` to a `ViewController` instance.

The action `trigger` of the user can be deal with in other groups of test cases such as the interaction between `AppDelegate` and `AppRouterSpy`.

ctwdtw committed:
add `AppRouter.register()` to eliminate anti-pattern of adding side effect on `init`

ctwdtw committed:
Move `sut.register()` into `makeSUT` to simplify tests so that we can read/fix test cases with common given block easier

the-yuyut reviewed (changes requested):
Made some changes on TragkForMemoryLeaks.

ctwdtw committed:
spike new test cases

ctwdtw committed:
inject mappings

ctwdtw committed:
add file, line parameters for `trackForMemroyLeaks`

Co-authored-by: Tsungyu Yu <42088295+ytyubox@users.noreply.github.com>

ctwdtw committed:
Merge branch 'tdd_appRouter' of https://github.com/ctwdtw/AppRouter into tdd_appRouter
